### PR TITLE
Feature/ner

### PR DESCRIPTION
**Pull request by YashSahsani**
## Description

1)Added ner endpoint.
2)Added test for the same.
3)Added necessary Documentation
4)Added Production docker file 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested? (Optional)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings
image
image

### CODE REVIEW

File: src/v1/service/Ner/ner.py [hunk 1]
```diff
@@ -1,12 +1,129 @@
...
+                help = "File to output NER results to.")
+    args = vars(ap.parse_args())
+
+    video_id = args['id']
+    video_transcript = get_transcript(video_id)
+    video_transcript = video_transcript.lower()
+    
+    ner_spacy = spacy_ner(video_transcript)
+    ner_nerd = nerd_ner(video_transcript)
+	"""
```
Comment (pandevim) [inline src/v1/service/Ner/ner.py:77]: We don't have to mention the whole code. Only a url would suffice.
```suggestion
# https://github.com/YouTubeNLP/NLP/tree/master/ner
```

File: README.md [hunk 2]
```diff
@@ -96,8 +96,8 @@ $ git clone https://github.com/YouTubeNLP/Backend.git && cd Backend/
 # For Local Testing
 $ pytest 
 # For Docker Testing
-$ docker build -t ynlp .
-$ docker --detach --name ynlp-backend-prod --publish 80:80 ynlp
+$ docker build -t ynlp .\
```
Comment (pandevim) [inline README.md:99]: Can you explain what the `.\` do here?

**Review by pandevim — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (YashSahsani) [reply README.md:99]: Its just for new line.

File: src/v1/service/Ner/ner.py [hunk 3]
```diff
@@ -1,12 +1,129 @@
...
+	    doc = nlp(document)
+	    results = [(ent.text, ent.label_) for ent in doc.ents]
+	    return results
+	# function to apply ner using nerd library
+	def nerd_ner(self,document):
+	    doc = ner.name(document, language='en_core_web_sm')
+	    results = [(ent.text, ent.label_) for ent in doc]
+	    return results
+	# function to save transcript
+	def save_file(self,document_name, transcript):
```
Comment (pandevim) [inline src/v1/service/Ner/ner.py:42]: Is saving a file doing any help in our process here?

File: src/v1/service/Ner/ner.py [hunk 4]
```diff
@@ -1,12 +1,129 @@
...
+	    file_name = args['file'] + '-' + document_name +'.txt'
+	    f = open(file_name, 'w+')
+	    for line in transcript:
+	        f.write(line + '\n')
+	    f.close()
+	# main method
+	def get_ner(self):
+	    video_transcript = self.get_transcript(self.video_id)	    
+	    ner_spacy = self.spacy_ner(video_transcript)
```
Comment (pandevim) [inline src/v1/service/Ner/ner.py:51]: If we're only using `nerd` why define `spacy` function and use it at all?
`
```suggestion

```

**Review by pandevim — CHANGES_REQUESTED** [review]
Also can you use a decent `videoId` for response screenshots?
Our service are primarily for political videos and not for music videos. So you can use the `videoId` suggested in the Backend [README.md](https://github.com/YouTubeNLP/Backend/blob/dev/README.md).

Also the structure of the response needs to change a little bit. Currently it's like this:
```json
[
  [ "entity", "label1" ],
  [ "entity", "label2" ],
  [ "entity", "label3" ]
]
```
We want it to be more compatible with the original schema which is decided. That is present in the [API Reference](https://app.swaggerhub.com/apis-docs/youtubenlp/backend/0.0.1#/default/get_ner__videoId_). That is:
```json
{
	"video_id": "video_id",
	"entities": [
		{
			"label1": "label",
			"text": [
				"string1",
				"string2",
				"string3"
			]
		},
		{
			"label2": "label",
			"text": [
				"string1",
				"string2",
				"string3"
			]
		},
		{
			"label3": "label",
			"text": [
				"string1",
				"string2",
				"string3"
			]
		}
	]
}
```
Also remember

File: README.md (hunk 2 above)
Comment (pandevim) [reply README.md:99]: Would you mind to explain how this is better than the previous one?

Comment (YashSahsani) [reply README.md:99]: if I don't put \ at the end of the line all the three commands are printed in one line.

**Comment by YashSahsani** [comment]
image

**Comment by YashSahsani** [comment]
image

**Review by pandevim — APPROVED** [review]
Nice work!

File: src/main.py [hunk 5]
```diff
@@ -110,9 +110,41 @@ async def controversial(video_id: str):
...
+                if(label_list[k] == label):
+                    label_num = k+1
+                    break
+            text_Available = False
+            for a in range(len(ners["entity"][label_num-1]["text"])):
+                    if(text == ners["entity"][label_num-1]["text"][a]):
+                        text_Available = True
+                        break
+            if(not text_Available):
+                ners["entity"][label_num-1]["text"].append(text)
```
Comment (pandevim) [inline src/main.py:143]: @YashSahsani 
My apologies this slipped my review before, but this logic is suppose to be inside the `Ner` service directory.

**Pull request merged**